### PR TITLE
[BEAM-6427]INTERSECT ALL is not compatible with SQL standard.

### PR DESCRIPTION
**Pull request by amaliujia**
Per SQL standard 1999 `7.12 3) iii) 3)`, 

For ALL case, 
```
 If INTERSECT is specified, then the number of duplicates of R that T contains is
the minimum of m and n. 
```

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

**Comment by amaliujia** [comment]
@akedin @kennknowles @apilloud

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamSetOperatorsTransforms.java [hunk 1]
```diff
@@ -80,8 +80,23 @@ public void processElement(ProcessContext ctx) {
...
             if (all) {
-              for (Row leftRow : leftRows) {
-                ctx.output(leftRow);
+              Iterator<Row> iter = leftRows.iterator();
+              int leftCount = 0;
+              int rightCount = 0;
+              while (iter.hasNext()) {
+                iter.next();
+                leftCount++;
+              }
```
Comment (akedin) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamSetOperatorsTransforms.java]: I think guava has `Iterators.size(iterator)`

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamSetOperatorsTransforms.java [hunk 2]
```diff
@@ -80,8 +80,23 @@ public void processElement(ProcessContext ctx) {
...
+              iter = rightRows.iterator();
+              while (iter.hasNext()) {
+                iter.next();
+                rightCount++;
+              }
+
+              // output MIN(m, n)
+              iter = (leftCount <= rightCount) ? leftRows.iterator() : rightRows.iterator();
+              while (iter.hasNext()) {
```
Comment (akedin) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamSetOperatorsTransforms.java]: Looking at this I am not sure I understand how SQL `INTERSECT` is supposed to work, can you add a brief explanation comment?

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamSetOperatorsTransforms.java [hunk 3]
```diff
@@ -17,6 +17,7 @@
  */
 package org.apache.beam.sdk.extensions.sql.impl.transform;
 
+import com.google.common.collect.Iterators;
```
Comment (akedin) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamSetOperatorsTransforms.java]: If I didn't miss anything we should use vendored guava: https://lists.apache.org/thread.html/6cdeb1f7cf6037c84a249268f641fa1caa52a721af017a1b643d60fa@%3Cdev.beam.apache.org%3E

Comment (kennknowles) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamSetOperatorsTransforms.java]: Yea, we depend on non-Vendored Guava as well, because `ValuesRel` requires an `ImmutableList` to a protected constructor. Other place we should not use it.

Comment (amaliujia) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamSetOperatorsTransforms.java]: using vendored Guava now.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamSetOperatorsTransforms.java (hunk 2 above)
Comment (akedin) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamSetOperatorsTransforms.java]: I see. What I was missing is how this whole thing works, i.e. that it runs `CoGBK` with keys and values being the whole row, producing a `CoGBKResult` with a collection of duplicates for each distinct row, which we then emit here.

Comment (akedin) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamSetOperatorsTransforms.java]: Can we instead potentially implement it with a combiner on both inputs, that counts the rows? I.e. `row -> KV<row,count> -> CoGBK -> emit min(count) number of times`? Would it be more optimal?

Comment (amaliujia) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamSetOperatorsTransforms.java]: If we cannot avoid shuffle, passing around value or counts will sometime not really impact performance. It really depends. I remember one problem faced by shuffle is too many small files. Because by the nature of shuffle, there will might be m * n connections where m is the number of mappers, n is the number of keys/small files on each mapper.  Some optimization on shuffle that might be done is compact files into larger files, or pipeline might enable external shuffle service. 

Basically I think one solution is not always be optimal all the time.
